yt_dlp/extractor/huya.py

```diff
@@ -0,0 +1,118 @@
+import base64
+import hashlib
+import json
+import random
+from html import unescape
+from urllib.parse import parse_qsl, urlencode
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    try_get
+)
+
+
+class HuyaIE(InfoExtractor):
+    _VALID_URL = r'(?x)^((?:http[s]?|fpt):)\/?\/(?:www\.|m\.|)huya\.com\/(?P<id>.*?)(?:\/|$)'
+    _TESTS = [{
+        'url': 'https://www.huya.com/kaerlol',
+        'info_dict': {
+            'id': 'kaerlol',
+            'ext': 'flv',
+            'title': '【1000攻击力】1A流戏命师',
+        },
+    }, {
+        'url': 'https://www.huya.com/573178',
+        'info_dict': {
+            'id': '573178',
+            'ext': 'flv',
+            'title': 'S6 全阵容细节教学',
+        },
+    }]
+    IE_NAME = 'huya'
+    IE_DESC = 'huya.com'
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id=video_id)
+
+        json_stream = self._search_regex(r'"stream": "([a-zA-Z0-9+=/]+)"', webpage, 'stream', default=None)
+
+        if not json_stream:
+            raise ExtractorError('Video is offline')
```

Comment:
```suggestion
            raise ExtractorError('Video is offline', expected=True)
```

Reply:
Ever thought about making it a separate exception class, say something like `ServiceDeclinedError`?

Reply:
It makes sense when designing the interface, but I don't see any advantage in adding it now